File: sdk/rust/src/messaging/stream.rs
/*
 * Copyright 2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * -----------------------------------------------------------------------------
 */
use messages::validator::Message;
use messages::validator::Message_MessageType;
use std;
use std::error::Error;
use std::sync::mpsc::Receiver;
use std::sync::mpsc::RecvError;
use std::time::Duration;

/// A Message Sender
///
/// A message
pub trait MessageSender {
    fn send(
        &self,
        destination: Message_MessageType,
        correlation_id: &str,
        contents: &[u8],
    ) -> Result<MessageFuture, SendError>;

    fn reply(
        &self,
        destination: Message_MessageType,
        correlation_id: &str,
        contents: &[u8],
    ) -> Result<(), SendError>;

    fn close(&mut self);
}

/// Result for a message received.
pub type MessageResult = Result<Message, ReceiveError>;

/// A message Receiver
pub type MessageReceiver = Receiver<MessageResult>;

/// A Message Connection
///
/// This denotes a connection which can create a MessageSender/Receiver pair.
pub trait MessageConnection<MS: MessageSender> {
    fn create(&self) -> (MS, MessageReceiver);
}

/// Errors that occur on sending a message.
#[derive(Debug)]
pub enum SendError {
    DisconnectedError,
    TimeoutError,
    UnknownError,
}

impl std::error::Error for SendError {
    fn description(&self) -> &str {
        match *self {
            SendError::DisconnectedError => "DisconnectedError",
            SendError::TimeoutError => "TimeoutError",
            SendError::UnknownError => "UnknownError",
        }
    }

    fn cause(&self) -> Option<&std::error::Error> {
        match *self {
            SendError::DisconnectedError => None,
            SendError::TimeoutError => None,
            SendError::UnknownError => None,
        }
    }
}

impl std::fmt::Display for SendError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match *self {
            SendError::DisconnectedError => write!(f, "DisconnectedError"),
            SendError::TimeoutError => write!(f, "TimeoutError"),
            SendError::UnknownError => write!(f, "UnknownError"),
        }
    }
}

/// Errors that occur on receiving a message.
#[derive(Debug, Clone)]
pub enum ReceiveError {
    TimeoutError,
    ChannelError(RecvError),
    DisconnectedError,
}

impl std::error::Error for ReceiveError {
    fn description(&self) -> &str {
        match *self {
            ReceiveError::TimeoutError => "TimeoutError",
            ReceiveError::ChannelError(ref err) => err.description(),
            ReceiveError::DisconnectedError => "DisconnectedError",
        }
    }

    fn cause(&self) -> Option<&std::error::Error> {
        match *self {
            ReceiveError::TimeoutError => None,
            ReceiveError::ChannelError(ref err) => Some(err),
            ReceiveError::DisconnectedError => None,
        }
    }
}

impl std::fmt::Display for ReceiveError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match *self {
            ReceiveError::TimeoutError => write!(f, "TimeoutError"),
            ReceiveError::ChannelError(ref err) => write!(f, "ChannelError: {}", err.description()),
            ReceiveError::DisconnectedError => write!(f, "DisconnectedError"),
        }
    }
}
/// MessageFuture is a promise for the reply to a sent message on connection.
pub struct MessageFuture {
    inner: Receiver<MessageResult>,
    result: Option<MessageResult>,
}

impl MessageFuture {
    pub fn new(inner: Receiver<MessageResult>) -> Self {
        MessageFuture {
            inner,
            result: None,
        }
    }

    pub fn get(&mut self) -> MessageResult {
        if let Some(ref result) = self.result {
            return result.clone();
        }

        match self.inner.recv() {
            Ok(result) => {
                self.result = Some(result.clone());
                result
            }
            Err(err) => Err(ReceiveError::ChannelError(err)),
        }
    }

    pub fn get_timeout(&mut self, timeout: Duration) -> MessageResult {
        if let Some(ref result) = self.result {
            return result.clone();
        }

        match self.inner.recv_timeout(timeout) {
            Ok(result) => {
                self.result = Some(result.clone());
                result
            }
            Err(_) => Err(ReceiveError::TimeoutError),
        }
    }
}

/// Queue for inbound messages, sent directly to this stream.

#[cfg(test)]
mod tests {

    use std::sync::mpsc::channel;
    use std::thread;

    use messages::validator::Message;
    use messages::validator::Message_MessageType;

    use super::MessageFuture;

    fn make_ping(correlation_id: &str) -> Message {
        let mut message = Message::new();
        message.set_message_type(Message_MessageType::PING_REQUEST);
        message.set_correlation_id(String::from(correlation_id));
        message.set_content(String::from("PING").into_bytes());

        message
    }

    #[test]
    fn future_get() {
        let (tx, rx) = channel();

        let mut fut = MessageFuture::new(rx);

        let t = thread::spawn(move || {
            tx.send(Ok(make_ping("my_test"))).unwrap();
        });

        let msg = fut.get().expect("Should have a message");

        t.join().unwrap();

        assert_eq!(msg, make_ping("my_test"));
    }
}

File: sdk/javascript/messaging/stream.js
/**
 * Copyright 2016 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ------------------------------------------------------------------------------
 */

'use strict'

const crypto = require('crypto')
const uuid = require('uuid/v4')
const zmq = require('zeromq')

const assert = require('assert')

const { Message } = require('../protobuf')
const Deferred = require('./deferred')
const { ValidatorConnectionError } = require('../processor/exceptions')

const _encodeMessage = (messageType, correlationId, content) => {
  assert(
    typeof messageType === 'number',
    `messageType must be a number; was ${messageType}`
  )
  assert(
    typeof correlationId === 'string',
    `correlationId must be a string; was ${correlationId}`
  )
  assert(
    content !== undefined || content !== null,
    'content must not be null or undefined'
  )
  assert(
    Buffer.isBuffer(content),
    `content must be a buffer; was ${
      content.constructor ? content.constructor.name : typeof content
    }`
  )

  return Message.encode({
    messageType,
    correlationId,
    content
  }).finish()
}

const _generateId = () =>
  crypto.createHash('sha256')
    .update(uuid())
    .digest('hex')

class Stream {
  constructor (url) {
    this._url = url

    this._initial_connection = true
  }

  connect (onConnectCb) {
    if (this._onConnectCb) {
      console.log(`Attempting to reconnect to ${this._url}`)
    }

    this._onConnectCb = onConnectCb

    this._futures = {}
    this._socket = zmq.socket('dealer')
    this._socket.setsockopt('identity', Buffer.from(uuid(), 'utf8'))
    this._socket.on('connect', () => {
      console.log(`Connected to ${this._url}`)
      onConnectCb()
    })
    this._socket.on('disconnect', (fd, endpoint) => this._handleDisconnect())
    this._socket.monitor(250, 0)
    this._socket.connect(this._url)

    this._initial_connection = false
  }

  close () {
    this._socket.setsockopt(zmq.ZMQ_LINGER, 0)
    this._socket.unmonitor()
    this._socket.close()
    this._socket = null
  }

  _handleDisconnect () {
    console.log(`Disconnected from ${this._url}`)
    this.close()
    Object.keys(this._futures).forEach((correlationId) => {
      this._futures[correlationId].reject(
        new ValidatorConnectionError('The connection to the validator was lost')
      )
    })

    this.connect(this._onConnectCb)
  }

  send (type, content) {
    if (this._socket) {
      const correlationId = _generateId()
      let deferred = new Deferred()
      this._futures[correlationId] = deferred

      try {
        this._socket.send(_encodeMessage(type, correlationId, content))
      } catch (e) {
        delete this._futures[correlationId]
        return Promise.reject(e)
      }

      return deferred.promise
        .then(result => {
          delete this._futures[correlationId]
          return result
        })
        .catch(err => {
          delete this._futures[correlationId]
          throw err
        })
    } else {
      let err = null
      if (this._initial_connection) {
        err = new Error('Must call `connect` before calling `send`')
      } else {
        err = new ValidatorConnectionError(
          'The connection to the validator was lost'
        )
      }

      return Promise.reject(err)
    }
  }

  sendBack (type, correlationId, content) {
    if (this._socket) {
      this._socket.send(_encodeMessage(type, correlationId, content))
    }
  }

  onReceive (cb) {
    this._socket.on('message', buffer => {
      let message = Message.decode(buffer)
      if (this._futures[message.correlationId]) {
        this._futures[message.correlationId].resolve(message.content)
      } else {
        process.nextTick(() => cb(message))
      }
    })
  }
}

module.exports = { Stream }
